packages/services/api/src/modules/preflight-script/providers/preflight-script.provider.ts

```diff
@@ -0,0 +1,215 @@
+      FROM "document_preflight_scripts"
+      WHERE "target_id" = ${targetSlug}
+      `);
+
+    return result && PreflightScriptModel.parse(result);
+  }
+
+  async createPreflightScript(
+    selector: TargetSelectorInput,
+    { sourceCode }: CreatePreflightScriptInput,
+  ) {
+    const res = validateSourceCode(sourceCode);
+    if (res) return res;
+
+    const [organizationId, projectId, targetId] = await Promise.all([
+      this.idTranslator.translateOrganizationId(selector),
+      this.idTranslator.translateProjectId(selector),
+      this.idTranslator.translateTargetId(selector),
+    ]);
+
+    await this.session.assertPerformAction({
+      action: 'laboratory:createPreflightScript',
+      organizationId,
+      params: {
+        organizationId,
+        projectId,
+        targetId,
+      },
+    });
+
+    const target = await this.storage.getTarget({
+      organizationId,
+      projectId,
+      targetId,
+    });
+
+    const currentUser = await this.session.getViewer();
+
+    const result = await this.pool.maybeOne(sql`/* createPreflightScript */
+      INSERT INTO "document_preflight_scripts" ( "source_code"
+                                               , "target_id"
+                                               , "created_by_user_id")
+      VALUES (${sourceCode},
+              ${targetId},
+              ${currentUser.id})
+      RETURNING
+          "id"
+          , "source_code" as "sourceCode"
+          , "target_id" as "targetId"
+          , "created_by_user_id" as "createdByUserId"
+          , to_json("created_at") as "createdAt"
+          , to_json("updated_at") as "updatedAt"
+      `);
+    const { data: preflightScript, error } = PreflightScriptModel.safeParse(result);
+
+    if (error) {
+      const { message } = fromZodError(error);
+      return {
+        error: {
+          __typename: 'PreflightScriptError' as const,
+          message,
+        },
+      };
+    }
+    return {
+      ok: {
+        __typename: 'PreflightScriptOkPayload' as const,
+        preflightScript,
+        updatedTarget: target,
+      },
+    };
+  }
+
+  async updatePreflightScript(selector: TargetSelectorInput, input: UpdatePreflightScriptInput) {
+    const res = validateSourceCode(input.sourceCode);
+    if (res) return res;
+
+    const [organizationId, projectId, targetId] = await Promise.all([
+      this.idTranslator.translateOrganizationId(selector),
+      this.idTranslator.translateProjectId(selector),
+      this.idTranslator.translateTargetId(selector),
+    ]);
+
+    await this.session.assertPerformAction({
+      action: 'laboratory:updatePreflightScript',
+      organizationId,
+      params: {
+        organizationId,
+        projectId,
+        targetId,
+      },
+    });
+
+    const target = await this.storage.getTarget({
+      organizationId,
+      projectId,
+      targetId,
+    });
+
+    const result = await this.pool.maybeOne(sql`/* updatePreflightScript */
```

Comment:
missing zod validation before inserting

Reply:
previously in `storage` we did parse after instertion, is something changed? https://github.com/graphql-hive/platform/blob/562a002b2ae02342972477610319229098b1f14f/packages/services/storage/src/index.ts#L598

Reply:
What you are referring to is to parse the data read from the database. There should still be input validation before inserting into the database.

Reply:
fixed https://github.com/graphql-hive/console/pull/5564#discussion_r1848045797